### ISSUE TITLE
Let the web UI adapt to the user's permissions

### ISSUE THREAD

**Issue by marians** [giantswarm/roadmap#551]
## User story

- _As a web interface user, I want the web interface to only show options which I can actually use, based on my permissions, so that I don't run into unexpected authorization errors._

### Milestone 1. Add support for read-only user

- [x] Cluster list (homepage)
  - [x] disable "Launch new cluster" button, adapt message in new cluster widget
  - [x] remove "Get started" button from cluster list item
  - [x] adapt cluster list empty state placeholder message
- [x] Cluster details overview
  - [x] disable cluster description editing, replace tooltip message
  - [x] remove "Add node pool" button
  - [x] remove install apps text
  - [x] remove create key pairs text
  - [x] disable cluster upgrade button, display permissions text
  - [x] disable release inspection modal upgrade options
  - [x] disable cluster labels editing, replace tooltip message, and remove add button
  - [x] disable switch to HA button, display permissions text
  - [x] remove "Get started" button
  - [x] add warning message to CLI guides
- [x] Cluster details > node pools
  - [x] disable "Add node pool" button, adapt empty state placeholder message
  - [x] disable node pool description editing, replace tooltip message
  - [x] disable node pool dots menu, add tooltip
  - [x] disable "Add node pool" button
  - [x] add warning message to CLI guides
- [x] Cluster details > client certificates
  - [x] add warning message to CLI guide
- [x] Cluster detail > apps > app details
  - [x] disable "Update..." button, add tooltip
  - [x] adapt app config messages and disable "Replace/Upload values" and "Replace/Upload secret values" buttons
  - [x] disable "Uninstall..." button
  - [x] add warning message to CLI guides
  - [x] *change "Install app" button into a link (for all users)
- [x] Cluster detail > ingress
  - [x] disable "Install ingress controller" button, adapt text
- [x] Cluster detail > action
  - [x] disable "Delete..." button, adapt text
  - [x] add warning message to CLI guide
- [x] Organizations overview
  - [x] disable "Add organization" button
  - [x] add warning message to CLI guide
- [x] Organization detail > general
  - [x] disable deletion button, adapt text
  - [x] adapt text for deletion blocked
  - [x] add warning message to CLI guide
- [x] Organization detail > access control
  - [x] remove add button for subjects
  - [x] remove icon for removing subjects
  - [x] add warning message to role details
  - [x] add warning message to CLI guide
- [x] Apps (catalogs page)
- [x] App detail
  - [x] *disable "Install in cluster" button (also disable if no clusters found), add tooltip
  - [x] add warning message to CLI guide
- [ ] Document the mechanism between Happa and RBAC in StoryBook, so that each team member can understand how all work.
- [x] https://github.com/giantswarm/roadmap/issues/602 

## Details

So far, the web interface works with a very simple permission model. It determines which organizations the user has access to. Then it fetches resources from these organization's namespaces and presents options for creation, modification or deletion of these resources.

However, role-based access control (RBAC) in Kubernetes allows much more fine-grained control with regard to which permissions each individual user should have. With RBAC, potentially, for each user and for each individual resource, each verb (like `get`, `update`, `delete` etc.) can be permitted or forbidden.

So in the web UI, before presenting e. g. a button to create a new node pool, we must first know whether the current user is actually permitted to create the intended resources in the intended namespace. More specifically, the button must only be shown if the user has `create` permissions for **all** necessary resources (depending on the provider and release; Azure: MachinePool, AzureMachinePool; AWS: MachineDeployment, AWSMachineDeployment). If at least one of these permissions is missing, the button must not be shown. All of this differentiation also needs to take the namespace (here: the org-namespace of the organization owning the workload cluster) into account.

**Permissions are dynamic.** Admins are free to bind new roles to users or groups at any time, or modify roles to add/remove permissions. We have to keep this in mind with our implementation, to balance the frequency of permission checks with the risk to either not show users all options they should have, or to show them too many options, potentially allowing to run into an authorization error after submission of an API request. Also we may prepare for the situation that continuously refreshing displayed information by fetching the resource again might at some point fail due to a new restriction in permissions.

### CLI Guides

CLI guides show commands/procedures to perform via `kubectl` or `kubectl gs` what is otherwise done in the web UI. To avoid users running into permission issues in the CLI, the easiest would be to only show a CLI guide if the user has all required permissions. Question: does that mean users are not learning about functions relevant to them? A question to potentially discuss with customers/users.

### Inform about what's inaccessible?

For users without read access to certain resources, the general question arises: should we inform users about something they are not seeing and that other users might see more, based on their permissions?

Specifications of the UI is [here](https://docs.google.com/presentation/d/1yCIv2B9Q9QmdN_s430r_NRqj0RobfQN7A3cAWioOsBk/edit#slide=id.gfc63257f33_0_0)

**Comment by marians** [comment]
An incomplete list of UI sections and how different permissions may play out.

## General

### Organization drop-down

## Home

### Cluster creation button

### Cluster listing

### Empty state

Currently we recommend to create a first cluster. If not allowed, we have to show a modified text.

(to be continued)

**Comment by marians** [comment]
## Re: Inform about what's inaccessible?

I see no simple rule of thumb to decide whether a UI element should simply be hidden when permissions are not sufficient, whether some information should be displayed instead. So here is some more verbose discussion:

**Are we puzzling users?** We don't want to. Whether we confuse our users by not showing some UI elements based on permissions depends a bit on how permissions are used in practice. If a user has the same permissions in all namespaces, clusters, cases, then some missing UI elements (e. g. no buttons for creating or deleting clusters, creating or deleting node pools) will be the norm for the user. They may not even know what they are missing. However, if permissions for one user vary a lot between organizations or clusters, that user might be looking for things they are used to in other places, unsuccessfully.

**Do we want users to know in general that they are missing something?** If they don't have access to a function or resource, it's not our goal to "create a desire" to access that function or resource. But we have an interest to help customers to have permissions configured right for all users. We can facilitate this in several ways, e. g.

- Showing the user which additional permission she is missing (ready to pass that info to an admin)
- Allow a user to request permission from an admin
- Allow an admin to impersonate a user
- Give an admin an overview of all the functions a user has and has not, based on their permissions

**Comment by puja108** [comment]
> For users without read access to certain resources, the general question arises: should we inform users about something they are not seeing and that other users might see more, based on their permissions?

I'm torn on this, on the one side this could help debugging, but on the other side I can imagine it feels not the nicest to know that your platform team (or whoever manages authz) limited you in certain ways. The debugging issue can be solved with other tools on the command line most probably even better. And in a way, maybe knowing what is limited right now is also a nice kind of transparency, and could foster conversations between the user and their auth team to negotiate higher level access, as they can see that it might help them solve something themselves instead of having to ask someone else.

TLDR: In summary I feel it could be beneficial to inform, but maybe not in a way that is always in your face. Like e.g. a section that you can expand to see what you are missing, but don't have to see it, if you're just doing your job and happy with it.

**Comment by puja108** [comment]
Exdeption to the above: Never show any names of things that a user does not have access to (namespaces, orgs, clusters, apps,...). This could have security implications.

**Comment by snizhana-dynnyk** [comment]
From the refinement:

### Informing about what's inaccessible
- Where it makes sense - we display all actions to all users but disable them for users that have insufficient permission to perform those actions together with a message explaining this.
- In some places (e.g. Get Started guide) we hide the action completely since it won't make sense for 'read-only' user to create key pairs for example.

### Next steps

- [ ] Make permissions auto-refreshable giantswarm/happa#2910
- [ ] Pairing on writing a spec

**Comment by PrashantR31** [comment]
I do not know how logging is at the moment, but it would be good to have the following logs:

**Comment by puja108** [comment]
@PrashantR31 as we are relying fully on Kubernetes as an API here, all those are limited by the logging feature sthat can be enabled there. The use cases you mention are mainly things that would be needed to add to audit logging, however AFAIK in K8s they'd not be as detailed on case 1, as the audit logs AFAIK do not show the contents of a patch.

Audit logging is definitely a valid user story, however, I'd separate it from this issue, as it pertains to everything users are doing in our API, no matter which UI is used.

Interesting side note here is if ownership of this is in Rainbow, which would bring it close to our discussions in today's product SWH about Rainbow owning certain extended components and configs around the API and I can definitely see the configuration of what gets into audit logs as part of that. We might have related audit logging stories already in the roadmap, but otherwise might make sense to track this in a new one. /cc @snizhana-dynnyk

**Comment by PrashantR31** [comment]
Alriiiight, thanks for the insight @puja108

**Comment by snizhana-dynnyk** [comment]
### Next steps

- @marians @kuosandys and @PrashantR31 organize a feedback session with customers
- continue working on spec
- slice the epic to user stories (refinement)

**Comment by marians** [comment]
## Spec update

We have meanwhile developed a spec for dealing with restricted permissions, talked about it with several customers, and got confirmation to proceed in the intended way.

Restricted permissions will be handled explicitly in the UI. This means that users will in general learn about the function they cannot access, and why. For example, a user who does not have all required permissions to create a cluster, will see a disabled "Launch new cluster" button on the front page, plus extra information. Likewise, Management API guides will be displayed completely, but will contain an according additional phrase to inform users about lacking permissions.

![image](https://user-images.githubusercontent.com/273727/143834680-e5bc2a10-6c91-4a37-8a83-67fb8f6e2946.png)

![image](https://user-images.githubusercontent.com/273727/143834737-d75295a4-31c4-4e70-acff-96c7789de3c2.png)

**Comment by snizhana-dynnyk** [comment]
##### Requirements of min permissions for a read-only user to access Web UI

- create self subject role reviews
- list organizations
- list/get clusters

**Comment by marians** [comment]
With a test user I ran into this error message on the organizations page.

<img width="342" alt="There was a problem loading" src="https://user-images.githubusercontent.com/273727/146550094-11639a90-7146-47ff-a57d-88166a1a6239.png">

The request URI: `https://happaapi.g8s..../apis/cluster.x-k8s.io/v1alpha3/clusters/`

This request is the attempt to list all clusters in all namespaces.

We should change the mechanics here to list clusters in the org-namespaces, plus in the default namespace. This way we can work without the need for the global permission.

**Comment by marians** [comment]
Similar to above, on the Apps tab under cluster details we are trying to load appcatalogentries from the cluster scope (not namespaced). We should restrict this to the namespace where we expect catalogs (`default`, plus the organization owning the cluster).

![image](https://user-images.githubusercontent.com/273727/146569711-8100b709-81ab-4faf-8010-76370ae786d4.png)

**Comment by marians** [comment]
Regarding Organizations > Access Control. When the user gets an empty result for (cluster) roles, this is what we show:

![image](https://user-images.githubusercontent.com/273727/149537636-e4b6f561-14a3-40c8-bc26-d83258ba0368.png)

There are several cases which we _could_ inform about and distinguish.

- User is not permitted to list ClusterRole resources (cluster scope).
- User is not permitted to list Role resources from the org-namespace.
- Despite permissions being granted, no items were returned.

Let's talk about this in the team and then come up with a better solution.

**Comment by marians** [comment]
We spoke about a detailed spec for the case where a user has access to the RoleBinding resources in the org-namespace, but not the ClusterRole resources. In this case, we want to show a limited view on the right hand side (no Managed-by, no description, no permissions).

Instead we show this disclaimer underneath the 🌐 Cluster role line:

    You need additional permissions to access ClusterRole details. Please talk to your administrator.

In this case the Permissions tab header will be disabled.

The logic for discovering ClusterRole resources has to be changed/extended, so that we first gather RoleBinding resources in the organization's namespace, and then collect the ClusterRole names pointed to in these bindings (`.roleRef.name`).

**Comment by marians** [comment]
In addition to the previous comment:

If the user has no permission to edit RoleBinding (but can read ClusterRole details):

    🚫  To edit subjects, you need additional permissions. Please talk to your administrator.

If the user has no permission to edit RoleBindings _and_ cannot read ClusterRoles:

    🚫  To edit subjects and to access cluster role details, you need additional permissions. Please talk to your administrator.

**Comment by PrashantR31** [comment]
Hello @marians and @kuosandys , a little help please. If you can review the following steps which I should go through, if they are relevant and in the proper order, so as to communicate properly before we move the issue to "Done"

Internal Release
1. Provide Internal Docs
2. Release on GS installation
3. Communicate to news-channel
4. If applicable, ask Engineers to test and add feedback to a new issue (Who are the engineers we need to ask to do that)


Customer Facing Release
1. Provide Customer facing Docs
2. Communicate to customers on their channels 
3. Demo with customers ( Specially with whom we presented the concept ) 
4. Release on Customer installation
5. Ask customers to provide feedback 

After all these actions are checked we can then move the issue to Done

Let me know please.
Thanks
Prashant

**Comment by kuosandys** [comment]
@PrashantR31 The changes within `happa` have mostly been released already, and running on all installations. We just haven't communicated the changes yet, and our "admin" users shouldn't see a difference as they don't have restricted permissions.

IMO I think we should work on the read-only cluster role first (or more accurately, set of cluster roles), before announcing the changes. As our milestone was targeted towards a specific type of user rather than the many combinations of permissions, this would make it easier for people to test and give feedback on.

**Comment by PrashantR31** [comment]
Perfect, let us work toward providing the UI to set Cluster role. Thanks @kuosandys.

**Comment by kuosandys** [comment]
@PrashantR31 We already have the access control UI for binding subjects to roles and cluster roles - is this what you're referring to? I think defining the cluster role(s) in the web UI would be a separate feature, and we haven't discussed this yet in the scope of this roadmap item.
